### ISSUE TITLE
SDF or URDF of Red Arm

### ISSUE THREAD

**Issue by reinzor** [daichi-yoshikawa/ahl_wbc#1]
Hi, I cannot find any launch files / URDF / SDF files of the Red arm, only blender and stl files.

Great packages!

**Comment by daichi-yoshikawa** [comment]
Thank you for referring to my packages !

Sorry, it's possible that I lost it (What did I do !?),
but I could find red_arm.yaml, which holds the parameters, in ahl_robot package.

I can recreate URDF file to enjoy the simulator in this weekend.
Can you wait till next week ?

Previously I made SDF file, but I wanted to make URDF instead of that.
This is good opportunity to do so :)

**Comment by reinzor** [comment]
Urdf file would be even better! Thanks. Yes I can wait till next week :)

-Rein

**Comment by reinzor** [comment]
Some other questions:
- Did you manage to implement your framework already on a real robot? 
- It would be nice give a simplistic example; the provided samples are already quit extensive with a GUI, a lot of joints etc. 

Thanks :).

I would really like to try to get this framework up and running on our new robotic arm. It does not have torque sensor though but we have current/torque control.

**Comment by daichi-yoshikawa** [comment]
> Did you manage to implement your framework already on a real robot?

Yes or No. I've tried to apply ahl_wbc to KUKA YouBot, which doesn't have torque sensor but current sensor. My challenge failed and it's probably because of the poor torque control (and time limitation. I had only one week to use the robot).

And  I'm trying to do re-challenge now. This is my private project, and I have to build a robot myself.
Or you can be the first guy who applies this package to a real robot.

> It would be nice give a simplistic example; the provided samples are already quit extensive with a GUI, a lot of joints etc.

Does it mean description on how to write config file, how to apply this software step by step and so on, assuming that people use simpler robot ?
If people need them, I write them.

> It does not have torque sensor though but we have current/torque control.

Does the torque control work very well ? If not, I recommend joint angle control instead of torque control.
ahl_wbc has API to convert desired torque to desired joint angles.
You can use it at least as the first step.

**Comment by daichi-yoshikawa** [comment]
Oh I guess that you want red arm URDF and simulation, because you're working on a robotic arm, right ?
So I'll add instruction how to apply this software to robotic arm by taking red arm as an example.
Is it good for you ?

**Comment by reinzor** [comment]
Yes, thanks a lot for all your effort!

-Rein

**Comment by daichi-yoshikawa** [comment]
I've created ahl_red_arm_description package, which contains red arm's URDF file.

Try the follows and you'll see red(orange?) arm.
Terminal 1: roslaunch ahl_red_arm_description red_arm.launch
Terminal 2: rosrun rviz rviz
On rviz, set parameters or add objects as an attached screenshot.
![screenshot from 2016-06-11 22 11 03](https://cloud.githubusercontent.com/assets/4362254/15985232/8d983a70-3021-11e6-9e33-d5982b4104f8.png)

I work on a control part from now on.

**Comment by daichi-yoshikawa** [comment]
I also implemented sample program for the red arm.

Step 1: roslaunch ahl_red_arm_description red_arm.launch
Step 2: roslaunch ahl_robot_samples red_arm.launch
Step 3: With a panel, select red_arm_sample->ahl_robot_controller and increase gains.
            (kp_task_pos -> 80 ~ 100, kp_task_ori -> 200 ~ 250)
Step 4: With the panel, select red_arm_sample->red_arm and move the arm.
            Add check sin_x/sin_y/sin_z, or change x_arm/y_arm/z_arm/roll_arm/pitch_arm/yaw_arm.

That's it !

**Comment by reinzor** [comment]
Thanks,

I will try tomorrow!

-Rein

**Comment by reinzor** [comment]
Thanks for all the effort; this will help me a lot! Another question:

Why did you decide to fork the gazebo_ros_pkgs (
https://github.com/daichi-yoshikawa/existing_packages/tree/master/gazebo_ros_pkgs)
and not use the original one (
https://github.com/ros-simulation/gazebo_ros_pkgs) and/or update them with
a fork on github?

-Rein

**Comment by daichi-yoshikawa** [comment]
That's a good point.

I should have update original one and thrown pull request.
At that time, my work was really experimental and I hesitated to update original one.

As for difference, I added APIs for torque control. Shared memory is utilized in it to transmit desired torques, current joint angles and so on.
Original one also have some APIs for torque control, but they are too slow because it uses
ROS service or message.

I'll add instructions to how to setup the simulation(Especially how to describe red_arm.yaml)
and how to use APIs. Wait till you see it.

**Comment by reinzor** [comment]
Ok, thanks for your feedback.

Another question: I see you developed the models in blender. Is it possible
for you to do a (video) tutorial of creating a URDF file from a blender
robot model? How to determine inertia's, export etc.

-Rein

On Wed, Jun 15, 2016 at 10:24 AM, Daichi Yoshikawa <notifications@github.com

**Comment by daichi-yoshikawa** [comment]
**After you read the follows, you might think you have to google a lot yourself. But I can add tutorials about them in this weekend(Now I can't, because I'm in crowded train in Tokyo... ). So if you wait till next week, you can focus on your business this week.**

> > Is it possible for you to do a (video) tutorial

Yes. As I mentioned above, I'll do it this weekend as possible. Of course, I'll keep updating it.

> > Creating URDF from blender

Blender cannot generate URDF file. I wrote it almost from scratch, referring to my previous work.

Blender can generate dae file and stl file. Basically I use dae file for rendering, because it includes material info. Stl, which doesn't contain material info, is utilized to define a bounding box to detect collision.

Both of dae and stl are based on the same blender model. I always make stl(bounding box) simpler to make collision detection simpler(Collision detection's computational cost is easy to get high !). Blender can do this vertex reduction.

> > How to determine inertia

As for red arm, almost no theoritical approach behind it. Because this red arm is imaginary. I was inspired by KUKA LWR, but it's original.

Here I describe the guideline to determine inertias. In some step, you need meshlab(sudo apt-get install meshlab).

**Case 1 : you use existing robot**
1. Google the robot and look for the parameters. For example, KUKA YouBot's inertia is on the web.
2. If you have CAD model, use meshlab to determine inertial matrix and center of gravity.
In many cases, it's assumed that mass distribution is unifrom.
3. If you don't have googlable info and CAD model, do the follows.
3-1. Estimate mass of each link firstly. This is perfectly up to you.
3-2. Approximate each link to simple shaped object like cylinder, and compute inertial matrix and center of gravity. Of course you assume that mass distribution is uniform.
4. If your simulation gets unstable, make mass or inertia matrix bigger and bigger. Sometimes, small mass object can't be simulated stably(Even if the mass value is real !!)

**Case 2 : you use imaginary robot**
Basically you can  take Step 3 and 4 of Case 1.

One more thing, meshlab will help you but it's kind of tricky when determining inertial matrix. Gazebo's website noted it but I also illustrate it in my coming tutorial.

**Comment by zainmehdi** [comment]
Daichi I need help with your files I am getting node not found error when running your files ...

**Comment by reinzor** [comment]
Node not found looks like a typo or compilation error. Please make sure everything is compiled properly first. If so, what nodes are available?

-Rein

**Comment by zainmehdi** [comment]
![screenshot from 2016-08-24 15 49 23](https://cloud.githubusercontent.com/assets/21212125/17920956/6850ca54-6a12-11e6-90cb-eedc6c24f018.png)
This is the scenario .... when I try to compile the code using catkin_make it doesnt compile completely with ls j8 error I dont know where I am going wrong as I am no expert with ros :(

**Comment by daichi-yoshikawa** [comment]
Problem you have is caused by the fact catkin_make is not completed as Reinzor said.
Did you finish catkin_make without any errors ?
Or you've not setup catkin workspace properly ?

**Comment by zainmehdi** [comment]
![screenshot from 2016-08-25 16 00 32](https://cloud.githubusercontent.com/assets/21212125/17959714/28eb82e8-6add-11e6-8375-19eab7debbc5.png)
Thanks alot for response
I am getting error as mentioned in the pic.
I am following the procedure you mentioned.
1) Made catkin workspace as mentioned in tutorial on ROS wiki
2) then I downloaded your files in src directory of workspace
3) i typed catkin_make which produced error. I dont know where I am going wrong. Can you hint anything ?

**Comment by daichi-yoshikawa** [comment]
Are you using ROS Jade ? Or Indigo ? Which version of gazebo are you using ?
Possibly you need to install something like "ros-jade-**".
Rein and I are daily ROS users, so it could be possible that we've already installed something required to use my packages unconsciously before and you haven't.

**Comment by reinzor** [comment]
You are not showing the error in your log, catkin build in parallel so please put the build error here. Anyway, what I think is the problem is that you are building against the wrong gazebo version. What version of gazebo are u using? What is your Ros distro?

-Rein

**Comment by daichi-yoshikawa** [comment]
Haha, Rein, the same questions from us lol.

**Comment by reinzor** [comment]
Haha, yes. Just saw your email when I answered :x

-Rein

**Comment by zainmehdi** [comment]
I am using ROS indigo with gazebo 7 . I will give it a try with ROS jade . and what do you mean by error log, and catkin build in parallel. Please ignore my lack of knowledge .

**Comment by reinzor** [comment]
I did not try it with indigo, jade did work with gazebo 5. 

The gazebo api went to some changes from 5 to 7 so thus can be the cause. I don't have access to a computer right now to test it under gazebo 7, indigo.

What I mean is that since catkin is making in parallel, it is not clear where it crashes from the  end of the log. The end of the log of catkin make -j 1 will show a better representation of the error since it uses only one thread for building. 

-Rein

**Comment by zainmehdi** [comment]
![screenshot from 2016-08-25 18 53 52](https://cloud.githubusercontent.com/assets/21212125/17964683/5ea808d0-6af5-11e6-9696-bd552e50c8b8.png)
![screenshot from 2016-08-25 18 54 03](https://cloud.githubusercontent.com/assets/21212125/17964684/5eab9c8e-6af5-11e6-937a-662a6ac3b93a.png)
I tried with -j 1 and I think there is something wrong with multi camera plug in file ...

**Comment by reinzor** [comment]
Hmm, still cannot see the error in your log, only depreciation warnings. Indeed this multicamera plugin. This is most likely because the api changed from g5 to g7. Trying ROS jade with g5 is the easiest way to just get this to work. 

Hover, @daichi, if we would like to make the code compatible with kinetic and g7, it would be nice to depend on the official ros-simulation repo for the ROS gazebo packages.

-Rein

**Comment by zainmehdi** [comment]
ok I will give it a try with jade and G7 and let you know my findings. thanks alot for your help :)

**Comment by zainmehdi** [comment]
Thanks alot to both of you guys. I gave it a try with jade and g5 and worked like a charm.  :)

**Comment by zainmehdi** [comment]
One more thing I am working with schunk 7 dof manipulator and want to apply the same task space control along with null space motion utilization in joint space. The thing you have implemented is advanced stuff and I am still learning in the beginner phase. How should I proceed ?. I am following gazebo tutorials on gazebosim. will they be good enough to do the kind of stuff you have done ? Thnx

**Comment by reinzor** [comment]
I think it is good to follow all gazebo tutorials so that you will get a hang of gazebo / ros. Those tutorials allow you to understand how to interact with the gazebo simulator with your own application. The whole body / task space control is such an application; however, a complex one ;).

**Comment by reinzor** [comment]
I think btw that this issue can be closed since it is going a little bit out of scope ..
